### PR TITLE
PLT-7111 PLT-7682 PLT-7684 PLT-7685 PLT-7683 Open roles

### PR DESCRIPTION
**Pull request by bwbush**
- Open-roles implementation in Runtime
- Safety checks for open roles
- Integration test for open roles
- Documentation for open roles

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [ ] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
        - [ ] Change made to Marlowe validator (@bwbush and @palas must be included as reviewers)
        - Review not required
        - [ ] Minor changes to non-critical code, documentation, nix derivations, configuration files, or scripts
        - [ ] Formatting, spelling, grammar, or reorganization
    - [x] Reviewer requested

## Summary by CodeRabbit

New Features:
- Added support for "open role" contracts, allowing parties not identified at the time of contract instantiation to participate later.
- Introduced the concept of "open roles" in Marlowe contracts, enhancing contract flexibility.
- Implemented safety analysis support for open-role contracts, improving contract security.

Refactor:
- Modified various functions and data types to handle role tokens and minting, improving code modularity and reusability.
- Updated the `MarloweScripts` data type and added new types and functions related to helper scripts, enhancing script management.

Tests:
- Added new test cases and scenarios for open roles and role tokens, ensuring the robustness of the new features.

Documentation:
- Added detailed documentation on open roles in Marlowe contracts, providing clear guidelines for users.

### CODE REVIEW

File: marlowe-apps/src/Language/Marlowe/Runtime/App/Build.hs [hunk 1]
```diff
@@ -55,7 +56,7 @@ buildCreation version' contract roles minUtxo metadata' =
   let roles' =
         if M.null roles
           then RoleTokensNone
-          else RoleTokensMint . mkMint . fmap (second (,Nothing)) . NE.fromList . M.toList $ roles
+          else RoleTokensMint . mkMint . fmap (second $ (,Nothing) . ToAddress) . NE.fromList . M.toList $ roles
```
[the remark this answers is not in the document]
Comment (palas) [reply marlowe-apps/src/Language/Marlowe/Runtime/App/Build.hs:59]: That is not right, the empty case was already handled before the PR, the result type of `roles'` is `RoleTokensConfig`, and the result of the `fmap` is a `NonEmpty (TokenName, (Address, Maybe RoleTokenMetadata))`, there is no `Destination` type anywhere

[the remark this answers is not in the document]
Comment (jhbertra) [reply marlowe-apps/src/Language/Marlowe/Runtime/App/Build.hs:59]: That's also incorrect. There is in fact there a `Destination` type and the `fmap` produces a `NonEmpty (TokenName, (Destination, Maybe RoleTokenMetadata))`. `ToAddress` is a data constructor of the `Destination` type which has the type signature `ToAddress :: Address -> Destination`.

[the remark this answers is not in the document]
Comment (palas) [reply marlowe-apps/src/Language/Marlowe/Runtime/App/Build.hs:59]: Huh, thanks

**Review by jhbertra — COMMENTED** [review]
I meant to comment this days ago... apparently I never submitted the review!

File: marlowe-runtime-web/src/Language/Marlowe/Runtime/Web/Types.hs [hunk 2]
```diff
@@ -830,14 +851,31 @@ instance ToSchema RolesConfig where
 data RoleTokenConfig
   = RoleTokenSimple Address
   | RoleTokenAdvanced Address TokenMetadata
+  | OpenRoleTokenSimple
+  | OpenRoleTokenAdvanced TokenMetadata
+  | ThreadRoleTokenSimple
+  | ThreadRoleTokenAdvanced TokenMetadata
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe-runtime-web/src/Language/Marlowe/Runtime/Web/Types.hs]: Fixed.

File: marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs [hunk 3]
```diff
@@ -121,7 +127,7 @@ createSpec = Hspec.describe "buildCreateConstraints" do
   Hspec.QuickCheck.prop "sends the minAda deposit to the marlowe output" \(SomeCreateArgs args) ->
     let result = extractMarloweAssets <$> runBuildCreateConstraints args
      in case version args of
-          MarloweV1 -> result === (Right $ Just $ Chain.Assets (minAda args) mempty)
+          MarloweV1 -> result === (Right $ Just $ Chain.Assets (max 2_000_000 $ minAda args) mempty)
```
[the remark this answers is not in the document]
Comment (bwbush) [reply marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/BuildConstraintsSpec.hs]: This had to do with a change in the workaround for a longstanding bug in cardano-api. I've revised it in 4a3341a5771f765bfde32c7876769ff62d0a29cb to make the test case cleaner to read.

**Pull request merged**